Lib/test/test_zipfile.py

```diff
@@ -2880,7 +2881,7 @@ def test_open(self):
             a, b, g = root.iterdir()
             with a.open() as strm:
                 data = strm.read()
-            assert data == b"content of a"
+            assert data == "content of a"
```

Comment:
This changes the return type of `Zipfile.Path.open` from bytes to string and could be backwards incompatible on backporting to 3.8

Reply:
Thanks for highlighting this concern. It's this concern that's kept me from fast-tracking this change. I'll try to enumerate the rationale and concerns and alternatives (to simply breaking it).

## Rationale

- `zipfile.Path` was intended to be compatible with a subset of `pathlib.Path`.
- Recent [efforts in importlib_resources](https://gitlab.com/python-devs/importlib_resources/issues/80) revealed that a `pathlib.Path` compatible implementation of `.open()` would be useful if not necessary.
- It would be beneficial for users for the two `Path` objects to have a consistent and compatible interface.
- The `.open()` probably should have been kept as a private (`._open()`) interface until it was needed.
- Users relying on `.open()` today can readily call `Path.root.open()` to get the current behavior.

## Concerns

- Users of Python 3.8 may be relying on the existing `.open()` implementation.
- The documentation [explicitly declares support for the existing behavior of `.open()`](https://docs.python.org/3/library/zipfile.html#zipfile.Path.open). If not for this concern, one might have been able to make the argument that `.open()` was an internal implementation detail.
- The "compatible subset" of pathlib.Path has previously been only informally defined, although there is an [ABC in importlib_resources](https://gitlab.com/python-devs/importlib_resources/-/blob/23e04a8c71dd199ef9e4029da3a6d9855a577937/importlib_resources/trees.py#L14) (slated for inclusion in Python 3.9) that aims to formalize that interface.
- There's no usage a user could supply that would work on both `.open()` methods as currently implemented. `pathlib.Path` requires `mode='rb'` to return bytes and `zipfile.Path` fails if 'b' is passed.
- The interface isn't just used by callers of `zipfile`, but also by consumers of APIs that return `zipfile.Path` (i.e. `importlib.metadata`).

## Alternatives

I'm considering several possible approaches.

- Simply break compatibility. The impact may be negligible given the limited exposure of `zipfile.Path`.
- Exclude this change in behavior from the (3.8) backport... and simply require all Python 3.8 users to deal with the incompatible interface.
- Provide a compatibility method like `_future_open` only on Python 3.8 that users could call to get the future behavior.
- Provide a compatible subclass, `FuturePath` only on Python 3.8 that would implement the alternate interface and would be supplied by `importlib.metadata` and others.
- Provide a backward-compatibility method like `_legacy_open` that provides the legacy behavior. That seems unnecessary given that `zipfile.Path.root.open` exists.

Regardless of how we proceed, there are some actions.

- [ ] Update docs to reflect change in API.

Reply:
After mulling this over throughout the week and capturing the rationale and concerns above, I'm leaning toward making this backward-incompatible change in Python 3.8.x, assuming you agree and think the concern is manageable. Here's how that would work:

- backport this code change as seen above
- docs would be updated to reflect the change happening in 3.8.(next).
- "what's new" would describe for users relying on `zipfile.Path.open` in Python 3.8.(next-1) to use `zipfile.Path.root.open` instead for easy, straightforward compatibility with the old behavior.

@ambv What do you think? Let me know if there's a better forum in which to discuss.